### PR TITLE
Add WithPartitionCount and WithDataVolume to Azure.Hosting.Azure.CosmosDB

### LINKED ISSUE #3295: Add support for cosmos emulator persistence
https://stackoverflow.com/questions/72466963/cosmosdb-emulator-linux-image-mount-volume

cc @Pilchie I can't find docs for this stuff.

### LINKED ISSUE #5136: [Aspire.Hosting.Azure.CosmosDB] Allow setting partition count and persistence
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

[Aspire.Hosting.Azure.CosmosDB](https://github.com/dotnet/aspire/tree/main/src/Aspire.Hosting.Azure.CosmosDB) component allows running CosmosDB as emulator which spins up the cosmosdb emulator docker image.

Now, the container is started with the default partition count (11) which makes the CosmosDB emulator slow to become available and take more memory. For me, 4 partitions is already enough.

Also, I believe the data is not persisted across runs.

### Describe the solution you'd like

The emulator supports a few environment variables that govern the behavior of the emulator:
| | Description |
| --- | --- |
| **`AZURE_COSMOS_EMULATOR_PARTITION_COUNT` *(Optional)*** | *Specify the number of partitions to use.* |
| **`AZURE_COSMOS_EMULATOR_ENABLE_DATA_PERSISTENCE` *(Optional)*** | *Enable data persistence between emulator runs.* |
| **`AZURE_COSMOS_EMULATOR_IP_ADDRESS_OVERRIDE` *(Optional)*** | *Override the emulator's default IP address.* |

I would like the Aspire.Hosting.Azure.CosmosDB component to support either extension methods to set these 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by hansmbakker**
I found #3295 and added the code to this PR.

Fixes https://github.com/dotnet/aspire/issues/5136
Fixes https://github.com/dotnet/aspire/issues/3295

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5157)

### CODE REVIEW

File: src/Aspire.Hosting.Azure.CosmosDB/AzureCosmosDBExtensions.cs [hunk 1]
```diff
@@ -124,6 +136,25 @@ public static IResourceBuilder<AzureCosmosDBEmulatorResource> WithGatewayPort(th
...
+    /// <param name="count">Desired partition count.</param>
+    /// <returns>Cosmos emulator resource builder.</returns>
+    /// <remarks>Not calling this method will result in the default of 25 partitions. The actual started partitions is always one more than specified.
+    /// See <a href="https://learn.microsoft.com/en-us/azure/cosmos-db/emulator-windows-arguments#change-the-number-of-default-containers">this documentation</a> about setting the partition count.
+    /// </remarks>
+    public static IResourceBuilder<AzureCosmosDBEmulatorResource> WithPartitionCount(this IResourceBuilder<AzureCosmosDBEmulatorResource> builder, int count)
+    {
+        if (count < 1 || count > 250)
+        {
+            throw new ArgumentOutOfRangeException(nameof(count), "Count must be between 1 and 250.");
```
Comment (martincostello) [inline src/Aspire.Hosting.Azure.CosmosDB/AzureCosmosDBExtensions.cs:152]: You could use the overload that provides the value that's out of range:

```suggestion
            throw new ArgumentOutOfRangeException(nameof(count), count, "Count must be between 1 and 250.");
```

**Comment by martincostello** [comment]
Sorry I don't know - it just looks like a flaky test.

**Comment by hansmbakker** [comment]
Ok, is it possible for you to rerun the failed builds? Or in general - what would you suggest as a next step for me?

**Comment by hansmbakker** [comment]
@martincostello all checks are green now 👍🏼

**Comment by hansmbakker** [comment]
@sebastienros some issues:
- `TestDistributedApplicationBuilder.CreateWithTestContainerRegistry(testOutputHelper)` results in an error with `netaspireci.azurecr.io/cosmosdb/linux/azure-cosmos-emulator:latest` not found. `TestDistributedApplicationBuilder.Create(options => { }, testOutputHelper)` is a workaround
- The cosmosdb emulator is extremely slow to start and will be a pain for all people / CI using this test as I guess it is not possible to keep it alive across tests (https://github.com/dotnet/aspire/issues/923 is still far away)

Is it an idea to add your request as a separate issue independent from this PR?

**Comment by sebastienros** [comment]
@hansmbakker we have a new testing API to wait for some text to be logged per resource. Here is an example for cosmosdb: https://github.com/dotnet/aspire/pull/5251/files/42d137900d1ec727e6ba09dcecf83ce9d8995a6b#diff-5051cd0c129b32d78bdb46214c5ec1f3db3b004e77ea64fe95791a1afc0ed520R144 

Note that this PR will be required as I had to add support for predicates instead of just texts, specifically for cosmosdb since the "Ready text" is used before as part of other logs.

As for the netaspireci.azurecr.io issue I will upload the image. In the meantime your suggestion with the `Create` call is fine.

For the other potential issue we haven't experienced it yet on the CI machines we use.

**Comment by sebastienros** [comment]
>As for the netaspireci.azurecr.io issue I will upload the image. In the meantime your suggestion with the Create call is fine.

Errata: we won't use netaspireci.azurecr.io since the image is on mcr.microsoft.com, so `Create` is the way to go. We do that in a few other places already. I also think that a new overload could be added to skip the first empty lambda.

**Comment by hansmbakker** [comment]
@sebastienros that looks useful but in my experience, just waiting for that text `Started` to be logged by the CosmosDB container is not enough.

I noticed that between the moment it logged `Started` and the moment that it starts returning valid responses to requests is still quite some time.

Other than that, it doesn't fix the point that the emulator takes about 2 minutes to start combined with the point that the Aspire tests won't reuse the container instance. In my opinion, that will result in very time-expensive tests.

**Comment by davidfowl** [comment]
cc @eerhardt

**Comment by davidfowl** [comment]
I would love to have this in Aspire 9 @eerhardt @sebastienros. 

cc @radical to see if he has any advice WRT time to boot the container for testing.

**Comment by radical** [comment]
`cosmos` emulator starts up faster with:
```csharp
                var cosmos = AppBuilder
                                .AddAzureCosmosDB("cosmos")
                                .RunAsEmulator(resource => resource.WithEnvironment("AZURE_COSMOS_EMULATOR_PARTITION_COUNT", "2"));
```

I'm adding this to the integration tests.

**Comment by davidfowl** [comment]
We should change our default too

File: src/Aspire.Hosting.Azure.CosmosDB/AzureCosmosDBExtensions.cs [hunk 2]
```diff
@@ -124,6 +136,25 @@ public static IResourceBuilder<AzureCosmosDBEmulatorResource> WithGatewayPort(th
...
+    /// See <a href="https://learn.microsoft.com/en-us/azure/cosmos-db/emulator-windows-arguments#change-the-number-of-default-containers">this documentation</a> about setting the partition count.
+    /// </remarks>
+    public static IResourceBuilder<AzureCosmosDBEmulatorResource> WithPartitionCount(this IResourceBuilder<AzureCosmosDBEmulatorResource> builder, int count)
+    {
+        if (count < 1 || count > 250)
+        {
+            throw new ArgumentOutOfRangeException(nameof(count), count, "Count must be between 1 and 250.");
+        }
+
+        return builder.WithEnvironment("AZURE_COSMOS_EMULATOR_PARTITION_COUNT", count.ToString(CultureInfo.InvariantCulture));
```
Comment (radical) [inline src/Aspire.Hosting.Azure.CosmosDB/AzureCosmosDBExtensions.cs:155]: Could you please add a test for this like in https://github.com/dotnet/aspire/pull/5641 ?

File: src/Aspire.Hosting.Azure.CosmosDB/AzureCosmosDBExtensions.cs [hunk 3]
```diff
@@ -110,6 +112,16 @@ public static IResourceBuilder<AzureCosmosDBResource> RunAsEmulator(this IResour
         return builder;
     }
 
+    /// <summary>
+    /// Adds a named volume for the data folder to an Azure Cosmos DB emulator resource.
+    /// </summary>
+    /// <param name="builder">The builder for the <see cref="AzureCosmosDBEmulatorResource"/>.</param>
+    /// <param name="name">The name of the volume. Defaults to an auto-generated name based on the application and resource names.</param>
+    /// <returns>A builder for the <see cref="AzureCosmosDBEmulatorResource"/>.</returns>
+    public static IResourceBuilder<AzureCosmosDBEmulatorResource> WithDataVolume(this IResourceBuilder<AzureCosmosDBEmulatorResource> builder, string? name = null)
```
Comment (radical) [inline src/Aspire.Hosting.Azure.CosmosDB/AzureCosmosDBExtensions.cs:121]: A test can be added for this like https://github.com/dotnet/aspire/blob/68ec8eadaa49bb170a19fba99b9a17462a821118/tests/Aspire.Hosting.Redis.Tests/RedisFunctionalTests.cs#L106-L189

**Comment by hansmbakker** [comment]
@radical thank you for the feedback! I'll try adding some tests, but will have to timebox it.

@davidfowl would it be possible for you to reach out internally regarding the reliability issues of the CosmosDB emulator image? As mentioned above, it is not only the slow startup of this image but rather the reliability issues that make working with this image a pain - e.g. when using it in integration tests ([with TestContainers on Azure Pipelines](https://github.com/testcontainers/testcontainers-dotnet/blob/62975bc135e0308a88ffbed84aad89fe9f5f1c3e/tests/Testcontainers.CosmosDb.Tests/CosmosDbContainerTest.cs#L17-L36)) it can lead to flaky tests.

Some issues my team ran into are:
- https://github.com/Azure/azure-cosmos-db-emulator-docker/issues/87
- https://github.com/Azure/azure-cosmos-db-emulator-docker/issues/45

Note that there are more reliability issues for other people:
- https://github.com/Azure/azure-cosmos-db-emulator-docker/issues/100
- https://github.com/Azure/azure-cosmos-db-emulator-docker/issues/104

File: src/Aspire.Hosting.Azure.CosmosDB/AzureCosmosDBExtensions.cs (hunk 2 above)
Comment (hansmbakker) [reply src/Aspire.Hosting.Azure.CosmosDB/AzureCosmosDBExtensions.cs:155]: done!

File: src/Aspire.Hosting.Azure.CosmosDB/AzureCosmosDBExtensions.cs (hunk 3 above)
Comment (hansmbakker) [reply src/Aspire.Hosting.Azure.CosmosDB/AzureCosmosDBExtensions.cs:121]: I added it - however I got issues with running multiple consecutive cosmosdb emulator instances and I don't have more time to spend on this unfortunately. Would any of you be able to have a look @radical @sebastienros?

**Comment by hansmbakker** [comment]
@radical @sebastienros I updated the PR according to the feedback I got and would like to hear what you think.

Also I see there are issues issues with running multiple consecutive cosmosdb emulator instances and I don't have more time to spend on this unfortunately. Would any of you be able to have a look?

cc @davidfowl

File: tests/Aspire.Hosting.Azure.CosmosDB.Tests/Aspire.Hosting.Azure.CosmosDB.Tests.csproj [hunk 4]
```diff
@@ -0,0 +1,14 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (eerhardt) [inline tests/Aspire.Hosting.Azure.CosmosDB.Tests/Aspire.Hosting.Azure.CosmosDB.Tests.csproj:1]: Do we need a separate assembly? Why not just include these in `Aspire.Hosting.Azure.Tests`?

**Comment by davidfowl** [comment]
I'd like to make sure we get this in for Aspire 9.

File: tests/Aspire.Hosting.Azure.CosmosDB.Tests/Aspire.Hosting.Azure.CosmosDB.Tests.csproj (hunk 4 above)
Comment (hansmbakker) [reply tests/Aspire.Hosting.Azure.CosmosDB.Tests/Aspire.Hosting.Azure.CosmosDB.Tests.csproj:1]: I followed the test projects of the other Aspire components - like the PostgreSQL one as @sebastienros gave as an example.

**Comment by hansmbakker** [comment]
@davidfowl could you please be more specific as to who you're asking for a next step?

I'm happy to contribute what I've done so far, but the cosmosdb emulator is not great to be used in automated tests, and unfortunately I cannot commit an unpredictable amount of time to this.

I asked for a hand on this, but nobody stepped in yet.

The branch is open to commits for maintainers of Aspire, so if this is important to have in Aspire 9, they are not blocked.

**Comment by davidfowl** [comment]
First, resolve the merge conflicts. Second, @eerhardt I think we should split out the azure resource tests, but I'm not sure it should be done as part of this PR.

@hansmbakker add the test here:

https://github.com/dotnet/aspire/blob/main/tests/Aspire.Hosting.Azure.Tests/AzureCosmosDBEmulatorFunctionalTests.cs

Delete the extra project. 

If this test is unreliable (which it might be because the cosmos emulator usually is), we'll disable the test (but keep the code for it).

**Comment by hansmbakker** [comment]
@davidfowl it seems ready to me 👍🏼 who can complete this?

**Review by davidfowl — APPROVED** [review]

**Comment by davidfowl** [comment]
Thanks @hansmbakker !

**Pull request merged**